src/collect/mod.rs

```diff
@@ -22,19 +22,59 @@ static PERF_DFLT_AUX_BUFSIZE: LazyLock<size_t> = LazyLock::new(|| {
 
 const PERF_DFLT_INITIAL_TRACE_BUFSIZE: size_t = 1024 * 1024; // 1MiB
 
-/// The interface offered by all trace collectors.
-pub trait TraceCollector: Send + Sync {
-    /// Obtain a `ThreadTraceCollector` for the current thread.
-    ///
-    /// A thread may obtain multiple `ThreadTraceCollector`s but must only collect a trace with one
-    /// at a time.
-    ///
-    /// FIXME: This API needs to be fixed:
-    /// https://github.com/ykjit/hwtracer/issues/101
+thread_local! {
+    /// When `Some` holds the `ThreadTraceCollector` that is collecting a trace of the current
+    /// thread.
+    static THREAD_TRACE_COLLECTOR: RefCell<Option<Box<dyn ThreadTraceCollector>>> = RefCell::new(None);
+}
+
+/// The private innards of a `TraceCollector`.
+pub(crate) trait TraceCollectorImpl: Send + Sync {
     unsafe fn thread_collector(&self) -> Box<dyn ThreadTraceCollector>;
 }
 
-pub trait ThreadTraceCollector {
+/// The public interface offered by all trace collectors.
+pub struct TraceCollector {
```

Comment:
It's OK if we do `Box` but if we expect `col_impl` to live longer than `TraceCollector` it might be worth doing:

```
struct TraceCollector<'a> {
   &'a dyn TraceCollectorImpl
}
```

I don't feels strongly about this though.

Reply:
Since it's the responsibility of the `TraceCollector` to both create and and drop `ThreadTraceCollector`s, I think an owned value makes more sense.

Reply:
Hold on, that's nonsense. My mistake.

If we were to store a reference to the `TraceCollectorImpl`, where would the owned value be stored? `TraceCollectorImpl` is deliberately private to the API consumer, so we would probably end up jumping through hoops to make such a design work.

Reply:
From a privacy perspective, `Box<TraceCollectorImpl>` and `&'a TraceCollectorImpl` are equivalent. Maybe I'm not understanding something?

Reply:
If you store a reference, then the owned value has to live somewhere else. My question is, where would this be?

Reply:
"if we expect col_impl to live longer than TraceCollector". Do we? It's your code, not mine :)

Reply:
`TraceCollector` and `TraceCollectorImpl` are part of the same conceptual object -- they are only separated to create a barrier between what's public and what's private. I expect them to be created together and to die together.

Reply:
Then what's the problem with `&'a`?

Reply:
Then the owned (private) inner would have to live outside of `TraceCollector` somewhere else. Probably the `TraceCollectorBuilder` would have to own it and give out a reference to it. This would tie the lifetime of the `TraceCollector` to the `TraceCollectorBuilder`. That would be a weird API.